src/main/scala/com/softwaremill/realworld/articles/comments/CommentsRepository.scala

```diff
@@ -52,56 +54,74 @@ class CommentsRepository(quill: Quill.Sqlite[SnakeCase]):
-        } yield (prIdOpt, arIdOpt)
-      }
+        pr <- queryProfile.join(pr => pr.userId == cr.authorId)
+        ar <- queryArticle.join(ar => cr.articleId == ar.articleId)
+      } yield (pr.userId, ar.articleId)
     )
-      .map(_.flatten.headOption)
+      .map(_.headOption)
 
-  def findComment(commentId: Int): Task[Option[Comment]] =
-    run(
-      for {
-        cr <- queryComment if cr.commentId == lift(commentId)
-        ar <- queryProfile.leftJoin(ar => ar.userId == cr.authorId)
-      } yield (cr, ar)
-    )
+  def findComment(commentId: Int, viewerId: Int): Task[Option[Comment]] = {
+    val commentRow: Quoted[Query[CommentRow]] = quote {
+      queryComment.filter(cr => cr.commentId == lift(commentId))
+    }
+
+    val commentQuery = buildCommentQueryWithFollowing(commentRow, viewerId)
+
+    run(commentQuery)
       .map(_.headOption)
-      .map(_.flatMap(comment))
+      .map(_.map(comment))
+  }
 
-  def findComments(articleId: Int): Task[List[Comment]] =
-    run(
-      for {
-        cr <- queryComment if cr.articleId == lift(articleId)
-        ar <- queryProfile.leftJoin(ar => ar.userId == cr.authorId)
-      } yield (cr, ar)
-    )
-      .map(_.flatMap(comment))
-
-  private def comment(tuple: (CommentRow, Option[ProfileRow])): Option[Comment] = {
-    val (cr, arOpt) = tuple
-
-    arOpt.map(ar =>
-      Comment(
-        id = cr.commentId,
-        createdAt = cr.createdAt,
-        updatedAt = cr.updatedAt,
-        body = cr.body,
-        author = CommentAuthor(
-          username = ar.username,
-          bio = ar.bio,
-          image = ar.image,
-          // TODO implement "following"
-          following = false
-        )
+  def findComments(articleId: Int, viewerIdOpt: Option[Int]): Task[List[Comment]] = {
+    val commentRow: Quoted[Query[CommentRow]] = quote {
+      queryComment.filter(cr => cr.articleId == lift(articleId))
+    }
+
+    val commentQuery = viewerIdOpt match
+      case Some(viewerId) => buildCommentQueryWithFollowing(commentRow, viewerId)
+      case None           => buildCommentQuery(commentRow)
+
+    run(commentQuery).map(_.map(comment))
+  }
+
+  private def comment(tuple: (CommentRow, ProfileRow, Boolean)): Comment = {
+    val (cr, pr, isFollowing) = tuple
+
+    Comment(
+      id = cr.commentId,
+      createdAt = cr.createdAt,
+      updatedAt = cr.updatedAt,
+      body = cr.body,
+      author = CommentAuthor(
+        username = pr.username,
+        bio = pr.bio,
+        image = pr.image,
+        following = isFollowing
       )
     )
   }
 
+  private def buildCommentQuery(crq: Quoted[Query[CommentRow]]) = {
+    quote {
+      for {
+        cr <- crq
+        pr <- queryProfile.join(ar => ar.userId == cr.authorId)
+      } yield (cr, pr, false)
+    }
+  }
+
+  private def buildCommentQueryWithFollowing(crq: Quoted[Query[CommentRow]], viewerId: Int) = {
+    quote {
+      for {
+        tuple <- buildCommentQuery(crq)
```

Comment:
**suggestion:** It's a bit unclear what's in this tuple. How about

```scala
for {
  (commentRow, profileRow, _) <- buildCommentQuery(crq)
  isFollowing = queryFollower
          .exists(f => (f.userId == profileRow.userId) && (f.followerId == lift(viewerId)))
  }
  yield (commentRow, profileRow, isFollowing)
```
(btw I'm also suggesting here to replace `.filter(...).nonEmpty` with `.exists(...)`

Reply:
`.exists(...)` method not exists on this query type :sweat_smile: